### PR TITLE
feat: Add versioning support for DynamoDB online store

### LINKED ISSUE #6163: Add feature view versioning support to DynamoDB online store
## Summary

Add versioned feature view read/write support to the `DynamoDB` online store, so that version-qualified feature references (e.g., `driver_stats@v2:trips_today`) resolve to the correct versioned data.

## Context

Feature view versioning was introduced in #6101. Currently, only the SQLite online store supports versioned reads/writes. All other online stores raise `VersionedOnlineReadNotSupported` when a versioned ref is used.

The implementation pattern (see `sqlite.py` for reference) involves:
1. **Write path**: When `enable_online_feature_view_versioning` is enabled, materialize data into a version-specific table/key namespace (e.g., `{project}_{fv_name}_v{version}`)
2. **Read path**: Route version-qualified lookups to the correct versioned table/key namespace

## Reference implementation

- SQLite online store: `sdk/python/feast/infra/online_stores/sqlite.py`
- Store file: `sdk/python/feast/infra/online_stores/dynamodb.py`

## Related

Part of #2728

### PR DESCRIPTION
**Pull request by rpathade**
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#code-style-and-linting
2. Run unit tests and ensure that they are passing: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#unit-tests
3. If your change introduces any API changes, make sure to update the integration tests here: https://github.com/feast-dev/feast/tree/master/sdk/python/tests
4. Make sure documentation is updated for your PR!
5. Make sure your commits are signed: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#signing-off-commits
6. Make sure your PR title follows conventional commits (e.g. fix: [Description of ...], feat: [Description of ...], chore: [Description of ...], refactor: [Description of ...])

-->
---                      
  What this PR does / why we need it

  Closes #6163 
                                                                                                               
  Feature view versioning was introduced in #6101, but version-qualified feature references (e.g.
  driver_stats@v2:trips_today) only worked with the SQLite online store. All other online stores raised        
  VersionedOnlineReadNotSupported when a versioned ref was used.                                             
                                                                                                               
  This PR adds versioned read/write support to the DynamoDB online store, following the same pattern as the  
  SQLite reference implementation:

  - Write path: when enable_online_feature_view_versioning is enabled, _get_table_name appends a _v{N} suffix  
  to the DynamoDB table name (e.g. project.driver_stats_v2), routing materialization to the correct versioned
  table                                                                                                        
  - Read path: version-qualified lookups (@v2) set projection.version_tag on the feature view before         
  online_read is called; _get_table_name picks this up and routes to the correct table. projection.version_tag 
  takes priority over current_version_number so explicit version requests are always honoured
  - Gate: _check_versioned_read_support in online_store.py now allows DynamoDBOnlineStore through, covering    
  both sync and async read paths                                                                               
   
  All existing read/write/update/teardown methods are covered by the single change to _get_table_name since    
  every method routes through it.                                                                            
                                                                                                               
  Which issue(s) this PR fixes                                                                               

  Part of #2728

  Checks

  - I've made sure the tests are passing.
  - My commits are signed off (git commit -s)
  - My PR title follows conventional commits format                                                            
   
  Testing Strategy                                                                                             
                                                                                                             
  - Unit tests — 6 new unit tests for _get_table_name covering: versioning disabled, projection.version_tag    
  priority, current_version_number fallback, version 0 edge case, no version info, custom template
  - Manual tests — end-to-end smoke test using moto (mocked DynamoDB) exercising write → read round trips      
  across v1/v2 isolation, fallback behaviour, and missing entity handling                                      
   
  Misc                                                                                                         
                                                                                                             
  The version suffix is composed before the table_name_template is applied, so custom table_name_template      
  values (e.g. "{project}__{table_name}") continue to work correctly with versioning enabled.
<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/feast-dev/feast/pull/6191" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

### CODE REVIEW

File: sdk/python/feast/infra/online_stores/online_store.py [hunk 1]
```diff
@@ -256,9 +256,10 @@ def get_online_features(
 
     def _check_versioned_read_support(self, grouped_refs):
         """Raise an error if versioned reads are attempted on unsupported stores."""
+        from feast.infra.online_stores.dynamodb import DynamoDBOnlineStore
```
Comment (ntkathole) [inline sdk/python/feast/infra/online_stores/online_store.py:259]: same here optional-dependency import

try:
    from feast.infra.online_stores.dynamodb import DynamoDBOnlineStore
except ImportError:
    DynamoDBOnlineStore = None

**Comment by ntkathole** [comment]
@rpathade need to handle teardown, feature_store.teardown() only contains the current FV with its current current_version_number. Previous version tables (e.g., project.driver_stats, project.driver_stats_v1) are not deleted when the current version is v2.

For DynamoDB this is higher severity, orphaned versioned tables will remain in AWS

**Comment by ntkathole** [comment]
Fixed via https://github.com/feast-dev/feast/pull/6257
